### PR TITLE
feat: OIDC Bearer token auth and documentation

### LINKED ISSUE #95: feat: OIDC Bearer Token Authentication
# OIDC Bearer Token Authentication

The server currently only supports HTTP Basic and Digest authentication, which require the client to send a username and password with every request. This is problematic when integrating with institutional identity providers (e.g. Keycloak) that issue short-lived OAuth2/OIDC access tokens. Users must either store long-lived credentials in the WebDAV config or resort to token-exchange proxies, both of which add operational complexity and weaken security posture.

## Solution

Add HTTP Bearer token authentication using OpenID Connect (OIDC). When a client sends `Authorization: Bearer <access_token>`, the server:

1. Verifies the JWT signature locally against the IdP's JWKS public keys (fetched at startup, no per-request network call).
2. Validates required claims: `iss`, `aud`, `azp`, `typ` ("Bearer"), `scope`, and `exp`.
3. Extracts the `preferred_username` claim and looks it up in `account_mapping`.
4. If the user is explicitly configured, that user's permissions apply. If not, the user inherits permissions from the `*oidc` sentinel template user.

Configuration uses a `*oidc` entry in `account_mapping` with a structured password format:

```
<oidc>#1#{issuer}#{jwks_uri}#{audience}#{client_id}#{algorithm}#{scope}
```

This follows the same sentinel-user convention as `*ldap` for consistency.

## Alternatives

- **Per-request token introspection** — Calling the IdP's introspection endpoint on every request provides immediate revocation but adds latency and load on the IdP. Local JWKS verification was preferred since access tokens are typically short-lived (5–15 min), making the revocation window acceptable for most deployments.
- **OAuth2 Proxy sidecar** — Running a separate OAuth2 Proxy in front of the WebDAV server adds operational overhead and breaks the single-binary deployment model.
- **Custom middleware/plugin** — A generic auth plugin system was considered but is out of scope for this feature; the OIDC integration fits naturally into the existing `DAVAuth` class.

## Additional context

- Implements the `*oidc` sentinel user pattern, mirroring the existing `*ldap` convention.
- OIDC is an optional extra (`pip install ASGIWebDAV[oidc]`) to avoid adding PyJWT/cryptography as hard dependencies.
- A single server can mix auth methods: Basic/Digest users, LDAP users, Bearer (OIDC) users, and anonymous access can coexist.
- Invalid Bearer tokens always return HTTP 401 (never silently degrade to anonymous).
- Basic auth against an OIDC-configured user explicitly fails — the `*oidc` password field is only used for JWT configuration.
- JWKS keys are fetched once at server startup; a restart is required to rotate keys.
- Full test suite included (247 lines) covering token verification, claim validation, user lookup, fallback behavior, and edge cases.

### PR DESCRIPTION
**Pull request by SilviaSWR**
# OIDC Bearer Token Authentication
# Checklist

- [x] I have read the [contribution guidelines](https://github.com/ASGI-WebDAV/ASGI-WebDAV/blob/main/CONTRIBUTING.md)
- [x] One Feature(issue) One PR
- [x] All commits in the PR will be merged into a single commit.
- [x] Add an entry in `changelog.en.md` if necessary? Don't forget to add your name and github profile link!
- [x] Add / update tests if necessary, Don't missing.
- [x] Add new / update outdated documentation

# Description

Add HTTP Bearer token authentication using OpenID Connect (OIDC). When a client sends `Authorization: Bearer <access_token>`, the server verifies the JWT locally against the IdP's JWKS public keys (fetched at startup, no per-request network call), validates required claims (`iss`, `aud`, `azp`, `typ`, `scope`, `exp`), extracts `preferred_username`, and resolves permissions from `account_mapping`.

## What's new

- **`HTTPOIDCAuth` class** (`asgi_webdav/auth.py`) — verifies JWTs locally against JWKS public keys. Eagerly fetches keys at startup; failure is fatal.
- **`DAVPasswordType.OIDC`** — new password type with format `<oidc>#1#issuer#jwks_uri#audience#client_id#algorithm#scope`.
- **`*oidc` sentinel user** — mirrors the `*ldap` convention. Configures the OIDC provider and serves as the default permission template for authenticated Bearer users not explicitly listed in `account_mapping`.
- **Bearer auth in `DAVAuth.pick_out_user()`** — handles `Authorization: Bearer` header, verifies token, looks up user permissions, falls back to `*oidc` template.
- **`WWW-Authenticate: Bearer`** appended to 401 challenge when OIDC is configured.

## Configuration

Add a `*oidc` entry to `account_mapping`:

```json
{
  "account_mapping": [
    {
      "username": "*oidc",
      "password": "<oidc>#1#https://idp.example.com/realms/PIC#https://idp.example.com/realms/PIC/protocol/openid-connect/certs#account#cosmohub-test#RS256#openid",
      "permissions": ["+^/$"]
    }
  ]
}
```

Install the optional dependency: `pip install ASGIWebDAV[oidc]`

## Key design decisions

- **Local verification** — JWTs are verified against JWKS public keys without calling the IdP per request. Revoked tokens remain valid until expiry (typically 5-15 min). This avoids IdP latency and load.
- **Fallback behavior** — Unknown Bearer users inherit `*oidc` template permissions. Invalid Bearer tokens always return 401 (never degrade to anonymous).
- **Basic auth rejection** — Basic auth against an OIDC-configured user explicitly fails; the `*oidc` password field is only used for JWT configuration.
- **Optional dependency** — PyJWT + cryptography are in a separate `[oidc]` extra to avoid bloating the core install.

## Files changed

| File | Change |
|------|--------|
| `asgi_webdav/auth.py` | `HTTPOIDCAuth` class, `DAVPasswordType.OIDC`, Bearer handling in `pick_out_user()` and `create_response_401()` |
| `requirements.d/oidc.txt` | New — `PyJWT>=2.11.0`, `cryptography` |
| `requirements.d/full.txt` | Added `-r oidc.txt` |
| `pyproject.toml` | Added `[oidc]` extra |
| `tests/conftest.py` | Skip tests if PyJWT not installed |
| `tests/test_auth_oidc.py` | 247-line test suite — token parsing, claim verification, user lookup, fallback, edge cases |
| `docs/guide/authentication.en.md` | Updated auth flow diagram, added Bearer auth section |
| `docs/guide/protect-your-password-in-the-config.en.md` | Full OIDC configuration reference, compatibility table |
| `docs/changelog.en.md` | v2.1.0 entry |

fixes #95

### CODE REVIEW

File: docs/guide/protect-your-password-in-the-config.en.md [hunk 1]
```diff
@@ -181,12 +181,132 @@ Specify the user DN pattern, with a `username` substitution field. Example:
...
+
+### Example configuration
+
+```json
+{
+  "account_mapping": [
+    {
+      "username": "*oidc",
+      "password": "<oidc>#1#https://idp.example.com/realms/PIC#https://idp.example.com/realms/institution/protocol/openid-connect/certs#account#cosmohub-test#RS256#openid",
+      "permissions": ["+^/$"]
```
Comment (rexzhang) [inline docs/guide/protect-your-password-in-the-config.en.md:258]: add a example that without `permissions` in config file

Comment (rexzhang) [reply docs/guide/protect-your-password-in-the-config.en.md:258]: all info(username and permissions) come from oauth maybe better in real deploy

**Comment by rexzhang** [comment]
Support OIDC is greate idea! Is it compatible with Authelia?

btw: I'll be 2-4 weeks before I have time to do a full review.

File: docs/guide/protect-your-password-in-the-config.en.md (hunk 1 above)
Comment (SilviaSWR) [reply docs/guide/protect-your-password-in-the-config.en.md:258]: Hey,

Implemented OIDC groups-based permission resolution. Permissions can now
come from the access_token's groups claim.

What changed:
- <oidc> password format: 8 → 9 fields, 9th is group_prefix (e.g.
  asgi-webdav_).
- HTTPOIDCAuth stores the prefix.
- New _extract_groups_permissions() filters token groups by prefix,
  strips prefix, returns permission list.
- pick_out_user now has 3-tier priority:
  1. User in config → config permissions (groups ignored)
  2. User not in config, has matching groups → groups as permissions
  3. User not in config, no matching groups → *oidc template fallback
- 6 new tests covering all priority paths, prefix filtering, empty
  groups, no groups claim.
- Updated docs.

**Comment by SilviaSWR** [comment]
This implementation is designed specifically for Keycloak. It relies on Keycloak-specific behavior, such as the `typ: Bearer` value in access tokens and the `preferred_username` claim. `Authelia` is not supported and would likely fail the `typ` validation.

File: docs/guide/protect-your-password-in-the-config.en.md (hunk 1 above)
Comment (rexzhang) [reply docs/guide/protect-your-password-in-the-config.en.md:258]: Please add a new example that without `permissions` in config file. If possible, please add the configuration information for the OIDC porvider(Keycloak) part.

**Comment by rexzhang** [comment]
> This implementation is designed specifically for Keycloak. It relies on Keycloak-specific behavior, such as the `typ: Bearer` value in access tokens and the `preferred_username` claim. `Authelia` is not supported and would likely fail the `typ` validation.

This is a real sad. I didn't use Keycloak, so please provide as much detailed documentation as possible on how to integrate it with Keycloak.

**Comment by SilviaSWR** [comment]
Sorry, I didn't explain myself clearly.

The feature was developed and tested using **Keycloak** backed by **FreeIPA**, but the implementation itself only relies on the standard **OpenID Connect** authentication flow and **OAuth 2.0** specifications. It does not depend on any Keycloak-specific APIs or features, so **in principle it should work with any standards-compliant OIDC provider**.

The only part that is provider-specific is the configuration of the issued access token. The application expects the access token to contain a set of standard claims (issuer, audience, scopes, username, etc.). These are mostly defined by the OpenID Connect and OAuth 2.0 specifications rather than by Keycloak itself.

Based on your comment, I realized this wasn't documented clearly enough, so I've added documentation describing the required access token claims and what each of them is used for. This should make it easier to configure other providers, such as Authelia, Authentik, Dex, or Zitadel, to be compatible with the application.

Since our environment uses Keycloak, I haven't been able to verify the configuration with Authelia myself. If you happen to configure OpenID Connect with Authelia following the documentation, I'd be interested to know whether everything works as expected. That would help confirm that the documentation is complete.

**Comment by SilviaSWR** [comment]
Hi @rexzhang, just following up on this PR. Have you had a chance to review it? I’d be happy to address any feedback or make any changes needed. Thanks!

**Comment by rexzhang** [comment]
Hi @SilviaSWR , I will review it during the October 1st holiday.
